src/coreclr/src/jit/optimizer.cpp

```diff
@@ -9271,55 +9273,94 @@ void Compiler::optRemoveRedundantZeroInits()
-                                        }
-                                    }
-                                }
+                        if (fieldRefCount != 0)
+                        {
+                            break;
+                        }
+
+                        // The local hasn't been referenced before this assignment.
+                        bool removedExplicitZeroInit = false;
+
+                        if (treeOp->gtOp2->IsIntegralConst(0))
+                        {
+                            if (!lclDsc->lvTracked)
+                            {
+                                bool bbInALoop  = (block->bbFlags & BBF_BACKWARD_JUMP) != 0;
+                                bool bbIsReturn = block->bbJumpKind == BBJ_RETURN;
 
-                                if (!removedExplicitZeroInit && treeOp->gtOp1->OperIs(GT_LCL_VAR) &&
-                                    (!canThrow || !lclDsc->lvLiveInOutOfHndlr))
+                                if (BitVecOps::IsMember(&bitVecTraits, zeroInitLocals, lclNum) ||
+                                    (lclDsc->lvIsStructField &&
+                                     BitVecOps::IsMember(&bitVecTraits, zeroInitLocals, lclDsc->lvParentLcl)) ||
+                                    !fgVarNeedsExplicitZeroInit(lclNum, bbInALoop, bbIsReturn))
                                 {
-                                    // If compMethodRequiresPInvokeFrame() returns true, lower may later
-                                    // insert a call to CORINFO_HELP_INIT_PINVOKE_FRAME which is a gc-safe point.
-                                    if (!lclDsc->HasGCPtr() ||
-                                        (!GetInterruptible() && !hasGCSafePoint && !compMethodRequiresPInvokeFrame()))
+                                    // We are guaranteed to have a zero initialization in the prolog or a
+                                    // dominating explicit zero initialization and the local hasn't been redefined
+                                    // between the prolog and this explicit zero initialization so the assignment
+                                    // can be safely removed.
+                                    if (tree == stmt->GetRootNode())
                                     {
-                                        // The local hasn't been used and won't be reported to the gc between
-                                        // the prolog and this explicit intialization. Therefore, it doesn't
-                                        // require zero initialization in the prolog.
-                                        lclDsc->lvHasExplicitInit = 1;
+                                        fgRemoveStmt(block, stmt);
+                                        removedExplicitZeroInit      = true;
+                                        lclDsc->lvSuppressedZeroInit = 1;
+                                        lclDsc->setLvRefCnt(lclDsc->lvRefCnt() - 1);
```

Comment:
We generally don't incrementally update ref counts anymore. Is this important?

Reply:
I added that in #37280. The idea is to detect cases where we make the struct dependently promoted because of a block assignment and then delete that block assignment. In that case we can make the field tracked and eliminate dead field assignments.

Reply:
I wonder if there is any guarantee that the count is > 0 at this point? I suppose this is running not long after the first call to `lvaMarkLocalVars` so the ref counts are probably still fairly accurate.

And the code to determine which locals to track is just a heuristic, so it's ok if it is based on incorrect ref count info.